### PR TITLE
nvgpu optimize reshape v3

### PR DESCRIPTION
**Pull request by fengleitian**
Another 3.7x compare to previous optimization. 
2,968us -> 779us

changes:
1.  remove dimension with size 1.
2. combine dimensions that's still in order after reorder
3. add 2D/3D shared memory tiled reshape. 
4. 2D reshape is 1.1x - 1.4x faster than cublas.
5. 3D reshape is 4 - 6x faster than previous optimization

many reshape can be use 2D/3D shared memory tiled reshape with cuda now, both are 4-6x faster.
for example: {1, 1, 1024, 256} -> {256, 1024, 1, 1} with input order {3, 2, 0, 1}
      become: {1024, 256} -> {256, 1024} with input order {1, 0}
which can use cublas matrix transpose.

for example: {256, 256, 3, 3} -> {3, 3, 256, 256} with input order {3, 2, 0, 1}
      become: {256, 256, 9} -> {9, 256, 256} with input order {2, 1, 0}
which can use 3D tiled reshape.

TODO: nd shared memory tiled reshape

```
Benchmarking ../ngraph-models/models/tensorflow/inference/resnet50/tf_function_ngraph_cluster_rn50_bs1.json, GPU backend, 10 iterations.
compile time: 7,043ms
11.5ms per iteration

---- Aggregate times per op type ----
Convolution     7,766us
Reshape         2,968us
BatchNorm         497us
Relu              334us
Add               202us
MaxPool            59us
Dot                34us
Sum                13us
Pad                10us
Divide              5us
Broadcast           5us
Result              1us
```

after:
```
enchmarking ../ngraph-models/models/tensorflow/inference/resnet50/tf_function_ngraph_cluster_rn50_bs1.json, GPU backend, 10 iterations.
compile time: 7,057ms
8ms per iteration

---- Aggregate times per op type ----
Convolution     6,565us
Reshape           779us
BatchNorm         463us
Relu              291us
Add               191us
MaxPool            53us
Dot                33us
Sum                11us
Pad                 8us
Divide              5us
Broadcast           4us
Result              2us
```

### CODE REVIEW

File: src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.cpp [hunk 1]
```diff
@@ -481,6 +481,122 @@ void runtime::gpu::CudaKernelBuilder::get_reshape_op(codegen::CodeWriter& writer
...
+                                                        runtime::gpu::GPUKernelArgs& args,
+                                                        const std::string& data_type,
+                                                        const std::vector<uint32_t>& order,
+                                                        const std::vector<uint32_t>& block_size)
+{
+    writer << "extern \"C\" __global__ void cuda_" << name << args.get_input_signature();
+    writer.block_begin();
+    {
+        writer << "__shared__ " << data_type << " tile[" << block_size[2] << "][" << block_size[1]
+               << "][33];\n";
```
Comment (csullivan) [inline src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.cpp]: Can we make the hardcoded `33` a parameter exposed in cuda_emitter?

File: src/ngraph/runtime/gpu/gpu_emitter.cpp [hunk 2]
```diff
@@ -540,39 +540,118 @@ namespace ngraph
...
+                }
+                // If there *is* a layout change in the 3D case, we do 3D tiled reshape.
+                else if (new_rank == 3)
+                {
+                    auto& cuda_emitter =
+                        external_function->get_primitive_emitter()->get_cuda_emitter();
+                    auto index = cuda_emitter->build_reshape_3d(
+                        {{args[0].get_type(), out[0].get_type()}}, new_arg_shape, new_input_order);
+
+                    writer << "gpu::invoke_primitive(ctx, " << index << ", ";
```
Comment (csullivan) [inline src/ngraph/runtime/gpu/gpu_emitter.cpp]: use `node_names`

File: src/ngraph/runtime/gpu/cuda_emitter.cpp [hunk 3]
```diff
@@ -653,6 +654,186 @@ size_t runtime::gpu::CUDAEmitter::build_reshape(const std::array<std::string, 2>
...
+    // 2. the other looks to see if this kernel is already in the primitive list
+
+    // check if the requested kernel is already an inserted primitive
+    size_t primitive_index = m_primitive_emitter->lookup(hash);
+    if (primitive_index != std::numeric_limits<size_t>::max())
+    {
+        return primitive_index;
+    }
+
+    // TODO: currently we set it to 64, will add tuning method later
```
[the remark this answers is not in the document]
Comment (csullivan) [reply src/ngraph/runtime/gpu/cuda_emitter.cpp]: Same with the comment in the other reshape emitter

File: src/ngraph/runtime/gpu/gpu_emitter.cpp [hunk 4]
```diff
@@ -540,39 +540,118 @@ namespace ngraph
...
-                // we can just copy.
+                //for a zero-size tensor, or change from 1^m shape to 1^n shape, just do a copy
                 if (!reshape->get_is_transpose() || result_shape_product < 2)
+                {
+                    kernel::emit_memcpyDtD(writer, out[0], args[0]);
+                    writer.block_end();
+                    return;
+                }
+
+                //combine inordered dimensons after reorder in shape, update output shape and input order
```
[the remark this answers is not in the document]
Comment (fengleitian) [reply src/ngraph/runtime/gpu/gpu_emitter.cpp]: good point.

File: src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.cpp (hunk 1 above)
Comment (fengleitian) [reply src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.cpp]: good catch, should be `block_size[0] + 1`, did. some test and forgot to change back.

File: src/ngraph/runtime/gpu/gpu_emitter.cpp (hunk 2 above)
Comment (fengleitian) [reply src/ngraph/runtime/gpu/gpu_emitter.cpp]: ya, right.

File: src/ngraph/runtime/gpu/gpu_emitter.cpp [hunk 5]
```diff
@@ -521,42 +521,120 @@ namespace ngraph
...
                 size_t result_shape_product = shape_size(result_shape);
 
-                // If there is no layout change or we are just going from 1^n to 1^m or a zero-size tensor,
-                // we can just copy.
-                if (same_layout || result_shape_product < 2)
+                //for a zero-size tensor, or change from 1^m shape to 1^n shape, just do a copy
+                if (!reshape->get_is_transpose() || result_shape_product < 2)
+                {
+                    kernel::emit_memcpyDtD(writer, out[0], args[0]);
+                    writer.block_end();
```
[the remark this answers is not in the document]
Comment (fengleitian) [reply src/ngraph/runtime/gpu/gpu_emitter.cpp]: changed.

**Comment by fengleitian** [comment]
Added tests to cover 3D/2D reshape.

**Pull request merged**